core/src/main/java/org/apache/iceberg/view/ViewMetadata.java

```diff
@@ -91,41 +111,325 @@ default ViewMetadata checkAndNormalize() {
+      // changes can be added
+      int highestFieldId = highestFieldId();
+      for (Schema schema : schemas) {
+        this.changes.add(new MetadataUpdate.AddSchema(schema, highestFieldId));
+      }
+
+      this.currentVersionId = newVersionId;
 
-    if (versionHistorySizeToKeep <= 0) {
-      LOG.warn(
-          "{} must be positive but was {}",
+      if (lastAddedVersionId != null && lastAddedVersionId == newVersionId) {
+        this.changes.add(new MetadataUpdate.SetCurrentViewVersion(LAST_ADDED));
+      } else {
+        this.changes.add(new MetadataUpdate.SetCurrentViewVersion(newVersionId));
+      }
+
+      return this;
+    }
+
+    private int highestFieldId() {
+      return schemas.stream().map(Schema::highestFieldId).max(Integer::compareTo).orElse(0);
+    }
+
+    public Builder setCurrentVersion(ViewVersion version) {
+      Schema schema = indexSchemas(schemas).get(version.schemaId());
+      int newSchemaId = addSchemaInternal(schema);
+      return setCurrentVersionId(
+          addVersionInternal(
+              ImmutableViewVersion.builder().from(version).schemaId(newSchemaId).build()));
+    }
+
+    private int reuseOrCreateNewViewVersionId(ViewVersion viewVersion) {
+      // if the view version already exists, use its id; otherwise use the highest id + 1
+      int newVersionId = viewVersion.versionId();
+      for (ViewVersion version : versions) {
+        if (version.equals(viewVersion)) {
+          return version.versionId();
+        } else if (version.versionId() >= newVersionId) {
+          newVersionId = viewVersion.versionId() + 1;
+        }
+      }
+
+      return newVersionId;
+    }
+
+    private int addVersionInternal(ViewVersion version) {
+      int newVersionId = reuseOrCreateNewViewVersionId(version);
+
+      if (versions.stream().anyMatch(v -> v.versionId() == newVersionId)) {
+        boolean isNewVersion =
+            lastAddedVersionId != null
+                && changes(MetadataUpdate.AddViewVersion.class)
+                    .anyMatch(added -> added.viewVersion().versionId() == newVersionId);
+        this.lastAddedVersionId = isNewVersion ? newVersionId : null;
+        return newVersionId;
+      }
+
+      ViewVersion newVersion;
+      if (newVersionId != version.versionId()) {
+        newVersion = ImmutableViewVersion.builder().from(version).versionId(newVersionId).build();
+      } else {
+        newVersion = version;
+      }
+
+      this.versions.add(newVersion);
+      this.changes.add(new MetadataUpdate.AddViewVersion(newVersion));
+      this.history.add(
+          ImmutableViewHistoryEntry.builder()
+              .timestampMillis(newVersion.timestampMillis())
+              .versionId(newVersion.versionId())
+              .build());
+      this.lastAddedVersionId = newVersionId;
+
+      return newVersionId;
+    }
+
+    private int addSchemaInternal(Schema schema) {
+      if (schema.schemaId() == -1) {
```

Comment:
I don't think this is handling a valid case. While the REST protocol uses -1 to signal the last added schema, it would be meaningless for a schema to use that ID.

Reply:
I agree that this wasn't really handling a valid case. This was rather an open TODO that we discussed internally. should be fixed now with your changes